packages/next/next-server/server/router.ts

```diff
@@ -41,4 +50,64 @@ export default class Router {
       }
     }
   }
+
+  async execute(
+    req: IncomingMessage,
+    res: ServerResponse,
+    parsedUrl: UrlWithParsedQuery
+  ): Promise<boolean> {
+    let parsedUrlUpdated = parsedUrl
+    let lastMatch:
+      | {
+          route: Route
+          result: RouteResult
+        }
+      | undefined
+    for (const route of this.routes) {
+      // If a redirect is matched we try to deeply resolve it so that you don't get
+      // double redirects. However this means we have to check until the type is not a redirect
+      // as otherwise there'd be an accidental redirect.
+      if (
+        lastMatch &&
+        lastMatch.route.type === 'redirect' &&
+        route.type !== 'redirect'
+      ) {
+        break
+      }
+
+      const params = route.match(parsedUrlUpdated.pathname)
+
+      // Check if the match function matched
+      if (params) {
+        const result = await route.fn(req, res, params, parsedUrlUpdated)
+        // The response was handled
+        if (result.finished) {
+          return true
+        }
+
+        lastMatch = {
+          route,
+          result,
+        }
+
+        if (result.pathname) {
+          parsedUrlUpdated.pathname = result.pathname
```

Comment:
Do we also need to support adding onto the `query` so we can maintain it for redirects?

Reply:
Discussed with @leo and we currently won't do that.